### PR TITLE
Update NIO to understand new directory stand-ins

### LINKED ISSUE #5203: New Google Cloud Storage folders not recognized by NIO
Of course, Google Cloud Storage does not actually have folders. However some tools create files with names ending in "/" to pretend that the folder exists.

NIO understands that 0-byte files that end in "/" actually mean "please pretend to the user that there is a folder with that name, even though Cloud Storage doesn't have folders", and complies (there is an option to turn that behavior off).

Recently we have seen non-empty files with the same intent (see [this bug report](https://github.com/broadinstitute/gatk/issues/5935)). We have to update NIO so as to keep track of the latest folder convention.

### PR DESCRIPTION
**Pull request by jean-philippe-martin**
The web interface has a "create folder" button, and that creates
a file ending in '/' that is used to pretend the folder exists.
The web interface maintains the illusion but of course this is still only a file,
so all clients (this one included) has to play along and pretend the directory
exists (or users get confused).

Those fake folders used to be 0-byte, but no longer. This pull request
updates our code so it just looks at the trailing slash, ignoring the size
of the file.

It also adds a test, and in order to write the test we had to add the ability
to create files whose name ends in '/', for test purpose only.

Fixes #5203

### CODE REVIEW

File: google-cloud-clients/google-cloud-contrib/google-cloud-nio/src/main/java/com/google/cloud/storage/contrib/nio/OptionAllowTrailingSlash.java [hunk 1]
```diff
@@ -0,0 +1,32 @@
+/*
+ * Copyright 2016 Google LLC
```
Comment (pmakani) [inline google-cloud-clients/google-cloud-contrib/google-cloud-nio/src/main/java/com/google/cloud/storage/contrib/nio/OptionAllowTrailingSlash.java]: small nits can you please update Copyright year ?

File: google-cloud-clients/google-cloud-contrib/google-cloud-nio/src/main/java/com/google/cloud/storage/contrib/nio/CloudStorageFileSystemProvider.java [hunk 2]
```diff
@@ -768,8 +769,14 @@ public void checkAccess(Path path, AccessMode... modes) throws IOException {
         }
         CloudStorageObjectAttributes ret;
         ret = new CloudStorageObjectAttributes(blobInfo);
-        // if size is 0 it could be a folder
-        if (ret.size() == 0 && cloudPath.seemsLikeADirectoryAndUsePseudoDirectories(storage)) {
+        // There exists a file with this name, yes. But should we pretend it's a directory?
```
Comment (pmakani) [inline google-cloud-clients/google-cloud-contrib/google-cloud-nio/src/main/java/com/google/cloud/storage/contrib/nio/CloudStorageFileSystemProvider.java]: please use multi line comments.

File: google-cloud-clients/google-cloud-contrib/google-cloud-nio/src/main/java/com/google/cloud/storage/contrib/nio/OptionAllowTrailingSlash.java (hunk 1 above)
Comment (jean-philippe-martin) [reply google-cloud-clients/google-cloud-contrib/google-cloud-nio/src/main/java/com/google/cloud/storage/contrib/nio/OptionAllowTrailingSlash.java]: Sure thing!

File: google-cloud-clients/google-cloud-contrib/google-cloud-nio/src/main/java/com/google/cloud/storage/contrib/nio/CloudStorageFileSystemProvider.java (hunk 2 above)
Comment (jean-philippe-martin) [reply google-cloud-clients/google-cloud-contrib/google-cloud-nio/src/main/java/com/google/cloud/storage/contrib/nio/CloudStorageFileSystemProvider.java]: OK!

**Comment by sduskis** [comment]
Please run `mvn com.coveo:fmt-maven-plugin:format` to fix the code formatting issues.

**Comment by jean-philippe-martin** [comment]
@sduskis done.

File: google-cloud-clients/google-cloud-contrib/google-cloud-nio/src/main/java/com/google/cloud/storage/contrib/nio/OptionAllowTrailingSlash.java (hunk 1 above)
Comment (chingor13) [reply google-cloud-clients/google-cloud-contrib/google-cloud-nio/src/main/java/com/google/cloud/storage/contrib/nio/OptionAllowTrailingSlash.java]: I think this was reversed - the new file should have the copyright year that it was introduced. So this one should be 2019 and the other one should stay 2016.

File: google-cloud-clients/google-cloud-contrib/google-cloud-nio/src/main/java/com/google/cloud/storage/contrib/nio/CloudStorageFileSystemProvider.java [hunk 3]
```diff
@@ -1,5 +1,5 @@
 /*
- * Copyright 2016 Google LLC
+ * Copyright 2019 Google LLC
```
Comment (chingor13) [inline google-cloud-clients/google-cloud-contrib/google-cloud-nio/src/main/java/com/google/cloud/storage/contrib/nio/CloudStorageFileSystemProvider.java]: This one should stay 2016

Comment (jean-philippe-martin) [reply google-cloud-clients/google-cloud-contrib/google-cloud-nio/src/main/java/com/google/cloud/storage/contrib/nio/CloudStorageFileSystemProvider.java]: You're right, sorry, I updated the wrong one. Should be fixed now.

**Comment by jean-philippe-martin** [comment]
I hope everyone had a good weekend! @chingor13 the code should be ready, let me know in case there's still a ball in my camp!

**Comment by jean-philippe-martin** [comment]
Thank you very much!

**Pull request merged**